CMakeLists.txt

```diff
@@ -416,86 +263,118 @@ if (S2N_INTERN_LIBCRYPTO)
+    try_compile(
+        IS_AVAILABLE
         ${CMAKE_BINARY_DIR}
-        SOURCES "${CMAKE_CURRENT_LIST_DIR}/tests/features/evp_md_ctx_set_pkey_ctx.c"
+        SOURCES "${CMAKE_CURRENT_LIST_DIR}/tests/features/${PROBE_NAME}.c"
         LINK_LIBRARIES ${LINK_LIB} ${OS_LIBS}
         CMAKE_FLAGS ${ADDITIONAL_FLAGS}
-)
+        COMPILE_DEFINITIONS -c ${PROBE_FLAGS}
+        ${ARGN}
+    )
 
-if (LIBCRYPTO_SUPPORTS_EVP_MD_CTX_SET_PKEY_CTX)
-    target_compile_options(${PROJECT_NAME} PUBLIC -DS2N_LIBCRYPTO_SUPPORTS_EVP_MD_CTX_SET_PKEY_CTX)
-endif()
+    # Set the result of the probe
+    feature_probe_result(${PROBE_NAME} ${IS_AVAILABLE})
 
-# Determine if KEM Kyber512 implementation from AWS-LC is available
-try_compile(
-        LIBCRYPTO_SUPPORTS_EVP_KEM_KYBER_512
-        ${CMAKE_BINARY_DIR}
-        SOURCES "${CMAKE_CURRENT_LIST_DIR}/tests/features/evp_kem_kyber_512.c"
-        LINK_LIBRARIES ${LINK_LIB} ${OS_LIBS}
-        COMPILE_DEFINITIONS "-Werror"
-)
+    # Make sure the variable is set in the parent scope
+    set(${PROBE_NAME} ${IS_AVAILABLE} PARENT_SCOPE)
 
-if(LIBCRYPTO_SUPPORTS_EVP_KEM_KYBER_512)
-    target_compile_options(${PROJECT_NAME} PUBLIC -DS2N_LIBCRYPTO_SUPPORTS_KYBER512)
-endif()
+    # Set the flags that we used for the probe
+    set(${PROBE_NAME}_FLAGS ${PROBE_FLAGS} PARENT_SCOPE)
+endfunction()
 
-# Determine if kTLS is supported
-try_compile(
-        PLATFORM_SUPPORTS_KTLS
-        ${CMAKE_BINARY_DIR}
-        SOURCES "${CMAKE_CURRENT_LIST_DIR}/tests/features/ktls.c"
-)
-if(PLATFORM_SUPPORTS_KTLS)
-    message(STATUS "feature probe: kTLS supported")
-    target_compile_options(${PROJECT_NAME} PUBLIC -DS2N_PLATFORM_SUPPORTS_KTLS)
-endif()
+# Iterate over all of the features and try to compile them
+FILE(GLOB FEATURE_SRCS "${CMAKE_CURRENT_LIST_DIR}/tests/features/*.c")
+list(SORT FEATURE_SRCS)
+foreach(file ${FEATURE_SRCS})
+    get_filename_component(feature_name ${file} NAME_WE)
+    feature_probe(${feature_name})
+endforeach()
 
-if (NOT DEFINED CMAKE_AR)
-	message(STATUS "CMAKE_AR undefined, setting to `ar` by default")
-	SET(CMAKE_AR ar)
-else()
-	message(STATUS "CMAKE_AR found: ${CMAKE_AR}")
+# FreeBSD need to link to execinfo explicitly
+if(NOT S2N_EXECINFO_AVAILABLE AND CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
+    feature_probe(S2N_EXECINFO_AVAILABLE LINK_LIBRARIES execinfo)
 endif()
 
-if (NOT DEFINED CMAKE_RANLIB)
-    message(STATUS "CMAKE_RANLIB undefined, setting to `ranlib` by default")
-    SET(CMAKE_RANLIB ranlib)
-else()
-    message(STATUS "CMAKE_RANLIB found: ${CMAKE_RANLIB}")
+# Stack traces are only available if execinfo is
+if (NOT S2N_EXECINFO_AVAILABLE)
+    set(S2N_STACKTRACE FALSE)
 endif()
+feature_probe_result(S2N_STACKTRACE ${S2N_STACKTRACE})
 
-if (NOT DEFINED CMAKE_OBJCOPY)
-    message(STATUS "CMAKE_OBJCOPY undefined, setting to `objcopy` by default")
-    SET(CMAKE_OBJCOPY objcopy)
-else()
-    message(STATUS "CMAKE_OBJCOPY found: ${CMAKE_OBJCOPY}")
+set(S2N_KYBER512R3_AVX2_BMI2 FALSE)
```

Comment:
In think so... Otherwise the value is undefined rather than false in the case where we don't set it to true. But I can try removing it and see if it causes issues.

Reply:
Yeah removing it causes the function argument to be empty:

```
$ cmake . -Bbuild -DBUILD_TESTING=on -DCMAKE_BUILD_TYPE=Debug -DS2N_NO_PQ=on
-- Detected CMAKE_SYSTEM_PROCESSOR as x86_64
-- Detected 64-Bit system
-- S2N_NO_PQ flag was detected - disabling PQ crypto
-- Found crypto: /nix/store/np58sm45gngld2nnqjq6p532j3v2kbfl-openssl-3.0.8/lib/libcrypto.so
-- LibCrypto Include Dir: /nix/store/gz31hsi6hdf0a2xzsznwfmv7vdw4ahcn-openssl-3.0.8-dev/include
-- LibCrypto Shared Lib:  /nix/store/np58sm45gngld2nnqjq6p532j3v2kbfl-openssl-3.0.8/lib/libcrypto.so
-- LibCrypto Static Lib:  crypto_STATIC_LIBRARY-NOTFOUND
-- Using libcrypto from the cmake path
-- CMAKE_AR found: /run/current-system/sw/bin/ar
-- CMAKE_RANLIB found: /run/current-system/sw/bin/ranlib
-- CMAKE_OBJCOPY found: /run/current-system/sw/bin/objcopy
-- feature S2N_CLONE_SUPPORTED: TRUE
-- feature S2N_CPUID_AVAILABLE: TRUE
-- feature S2N_EXECINFO_AVAILABLE: TRUE
-- feature S2N_FALL_THROUGH_SUPPORTED: TRUE
-- feature S2N_FEATURES_AVAILABLE: TRUE
-- feature S2N_KYBER512R3_AVX2_BMI2_SUPPORTED: TRUE
-- feature S2N_KYBER512R3_M256_INTRINSICS_SUPPORTED: TRUE
-- feature S2N_LIBCRYPTO_SUPPORTS_EVP_MD5_SHA1_HASH: TRUE
-- feature S2N_LIBCRYPTO_SUPPORTS_EVP_MD_CTX_SET_PKEY_CTX: TRUE
-- feature S2N_LIBCRYPTO_SUPPORTS_EVP_RC4: TRUE
-- feature S2N_LIBCRYPTO_SUPPORTS_KYBER512: FALSE
-- feature S2N_MADVISE_SUPPORTED: TRUE
-- feature S2N_MINHERIT_SUPPORTED: FALSE
-- feature S2N_PLATFORM_SUPPORTS_KTLS: TRUE
-- feature S2N___RESTRICT__SUPPORTED: TRUE
-- feature S2N_STACKTRACE: TRUE
CMake Error at CMakeLists.txt:376 (feature_probe_result):
  feature_probe_result Function invoked with incorrect arguments for function
  named: feature_probe_result
```